### ISSUE TITLE
no-redundant-callback for pure functions

### ISSUE THREAD

**Issue by aminya** [eslint/eslint#14065]
**Please describe what the rule should do:**
When a function (or lambda) is pure, there is no need to wrap it in another callback for passing it to other functions.

**A detailed example:**
Assume a function accepts functions as its arguments:
```ts
denounce(fun: Function, num: number)
```

Now, when using it, there should not be redundant callbacks to generate a trivial wrapper callback instead of passing the function directly:

Good
```js
// pure
function myCallback() {
	return 1
}

const debounceMyCallback = debounce(myCallback, 100)
```

Bad
```js
// pure
function myCallback() {
	return 1
}

const debounceMyCallback = denounce(()=> myCallback(), 100) // arrow callback is redundant.
```

**What new ECMAScript feature does this rule relate to?**
Arrow functions

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
-------------------

Good
```js
// pure
function myCallback(x, y) {
	return x + y
}

const debounceMyCallback = debounce(myCallback, 100)
```

Bad
```js
// pure
function myCallback(x, y) {
	return x + y
}

const debounceMyCallback = denounce((x, y)=> myCallback(x, y), 100) // arrow callback is redundant.
```

--------------------
Nothing should change in impure functions.
Good
```js
// impure 
function myCallback(x, y) {
	return x + y + z
}


const debounceMyCallback = denounce((x, y)=> myCallback(x, y, 1), 100) // OK
```
----------------------

The same thing can be done for pure lambdas, but I assume that the static analysis for lambdas is harder.
Bad
```js
// pure
const myCallback = (x, y) => {
	return x + y
}

const debounceMyCallback = denounce((x, y)=> myCallback(x, y), 100) // arrow callback is redundant.
```
--------------------------

**Why should this rule be included in ESLint (instead of a plugin)?**

- Doing this results in unnecessary lambda definition on the heap, which can result in slow garbage collection and thus lags in the code. 
- It also increases the size of the code with no benefits.
- The overhead of the wrapper callback can slow down the runtime (e.g. indirections in function pointers)

**Are you willing to submit a pull request to implement this rule?**

Sure, I can help.

**Comment by mdjermanovic** [comment]
Hi @aminya,

Thanks for the new rule proposal, but it doesn't qualify per [our rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), as it's a stylistic rule not related to an ECMAScript feature added within the last 12 months, so I'm closing this issue.

Regarding the proposal, I'm not sure if pure/impure makes a difference, but the number of passed vs used arguments can, and there's a rule that actually requires the opposite in some cases: [unicorn/no-array-callback-reference](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-array-callback-reference.md).

**Comment by aminya** [comment]
OK. This can become a third-party plugin.

> Regarding the proposal, I'm not sure if pure/impure makes a difference, but the number of passed vs used arguments can

When the function is impure, the additional scope of the lambda is needed to get those impure variables in. In other cases, the lambda is redundant and just results in performance drawbacks.

> there's a rule that actually requires the opposite in some cases: unicorn/no-array-callback-reference.

That's an opinionated code rule, which is not based on the programming science which cares about the performance of the code.
